proposals/generic-ref-struct.md

```diff
@@ -0,0 +1,424 @@
+
+- A `ref struct` that implements an `interface` must provide implementations for all methods, even those with a default implementation
+- A DIM invocation on a `ref struct` will result in a runtime exception
+
+The ability to implement interfaces does not change the rules around conversions though. A `ref struct` cannot be boxed, conversion from a `struct` to an `interface` has an implicit box associated with it hence conversion from `ref struct` to its implemented interfaces remains an error: 
+
+```c#
+ref struct S : IUtil { ... }
+
+S local = ...;
+IUtil util = local; // Error: a ref struct cannot be boxed 
+```
+
+The interfaces implemented by a `ref struct` can be used to satisfy constraints on a generic type parameter. Invocation of the interface methods on the type parameters is allowed because such invocations use the `.constrained` prefix for execution which eliminates boxing when the underlying type is a `struct`. 
+
+```c# 
+void Write<T>(T value)
+    where T : ref struct, ISpanFormattable
+{
+    var buffer = stackalloc char[100];
+
+    // Constrained interface call which does not box
+    if (value.TryFormat(buffer, out var written, default, null))
+    {
+        this.Write(buffer);
+    }
+}
+```
+
+### delegates
+The features described thus far allow developers the flexibility to opt existing generic types into `where T : ref struct` where legal. This though requires developers modify existing types and deal with any incompatibilities that may occur. Essentially it is not automatic, it requires work. 
+
+There is one class of types for which no human inpsection is needed: delegates. Delegates represent types that are signature only, there is no implementation to consider. That means the compiler itself can infer whether a type parameter could be validly `where T : ref struct` or not. It can do so by examining the places where type parameters are used and validating if the constraints are compatible with `where T : ref struct`. For example: 
+
+```c#
+// T is never used inside another generic hence implicitly supports ref struct
+delegate T Delegate1<T>(); 
+delegate void Delegate2<T>(T p);
+delegate void Deleaget3<T>(ref T p);
+
+// T is used in a generic param that does not support ref struct hence it does not 
+// support it
+delegate void Delegate4<T>(List<T> list);
+```
+
+The compiler will automatically mark type parameters on such `delegate` declarations as having the `where T : ref struct` annotation. This will eexpediate the adoption of this feature and not force a lot of extra annotations by developers. Consider that this will automatically make all known `Action` and `Func` delegates available for `ref struct` usage.
```

Comment:
```suggestion
The compiler will automatically mark type parameters on such `delegate` declarations as having the `where T : ref struct` annotation. This will expedite the adoption of this feature and not force a lot of extra annotations by developers. Consider that this will automatically make all known `Action` and `Func` delegates available for `ref struct` usage.
```